core/domain/exp_services.py

```diff
@@ -3590,8 +3592,32 @@ def update_logged_out_user_progress(
     checkpoint_url_model.put()
 
 
+@overload
+def sync_logged_out_learner_checkpoint_progress_with_current_exp_version(
+    exploration_id: str, unique_progress_url_id: str, *, strict: Literal[True]
+) -> exp_domain.TransientCheckpointUrl: ...
+
+
+@overload
+def sync_logged_out_learner_checkpoint_progress_with_current_exp_version(
+    exploration_id: str, unique_progress_url_id: str,
+) -> Optional[exp_domain.TransientCheckpointUrl]: ...
+
+
+@overload
 def sync_logged_out_learner_checkpoint_progress_with_current_exp_version(
-    exploration_id: str, unique_progress_url_id: str
+    exploration_id: str, unique_progress_url_id: str, *, strict: Literal[False]
+) -> Optional[exp_domain.TransientCheckpointUrl]: ...
+
+
+@overload
+def sync_logged_out_learner_checkpoint_progress_with_current_exp_version(
+    exploration_id: str, unique_progress_url_id: str, *, strict: bool
+) -> Optional[exp_domain.TransientCheckpointUrl]: ...
```

Comment:
Why is this overload needed?

Reply:
> Why is this overload needed?

Because at this point in the code https://github.com/oppia/oppia/blob/d72b38601acdb90b933d2c5cb57ee3501e3a285b/core/controllers/reader.py#L320 we are fetching value from the function `sync_logged_out_learner_checkpoint_progress_with_current_exp_version`, and previously this method can return None value as well so `synced_exp_user_data` can be None and further in the code we are accessing some attributes on this `synced_exp_user_data` variable. So if `synced_exp_user_data` is None then a python error will occur `attribute is not defined on None object`. So to rule out the None possibility here and if None occurs somehow here, then to throw a more explicit error instead of python we defined strict paradigm here.

Reply:
Yea, for that, I don't think this specific `overload` is necessary. 

It is a redundant to the overload above it

Reply:
Done!, removed it!